streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java

```diff
@@ -35,18 +35,19 @@
 import java.util.TreeMap;
 
 /**
- * Manages the {@link Segment}s that are used by the {@link RocksDBSegmentedBytesStore}
+ * Manages the {@link KeyValueSegment}s or {@link TimestampedSegment} that are used by the {@link RocksDBSegmentedBytesStore}
+ * and {@link RocksDBTimestampedSegmentedBytesStore}.
  */
-class Segments {
-    private static final Logger log = LoggerFactory.getLogger(Segments.class);
+abstract class AbstractSegments<S extends Segment> {
+    private static final Logger log = LoggerFactory.getLogger(AbstractSegments.class);
 
-    private final TreeMap<Long, Segment> segments = new TreeMap<>();
-    private final String name;
+    final TreeMap<Long, S> segments = new TreeMap<>();
+    final String name;
```

Comment:
Thanks to the flat package structure, this change actually exposes these fields to 66 other classes, all of our other internal state-related code. This is actually pretty risky, since `segments` in particular is mutable.

Your intent is to broaden the scope to only the subclasses, but Java has no visibility modifier to do that. (`protected` seems to be the one, but its visibility is actually broader than package-protected)

To enforce the desired visibility restriction, we would have to create a new package, `org.apache.kafka.streams.state.internals.segments`, and locate the abstract class and both implementations in it. Then the fields would be visible only to the implementing classes, as desired.

Personally, I think the mutable `segments` field justifies this, as future internal state-store code changes might accidentally think it's ok to touch `segments`, directly, leading to subtle bugs.

Reply:
I don't see this danger. Note, that all changes must be approved by a committer and thus, I am sure this would be caught (it's not a project for which somebody will limited knowledge can easily push changes). If you insist on a different package structure I am also fine with it, even if I don't see a bid advantage introducing many small packages.

\cc @guozhangwang

Reply:
In general, there may not be a big advantage to many small packages, but in this particular case, it happens to be the only way to safely protect what is intended to be private state.

Personally, I wouldn't lean too heavily on the eternal vigilence of reviewers to catch subtle bugs when there's an easy way to prevent them structurally. I certainly wouldn't trust _myself_ to just remember that this field, although it's made accessible to the whole package, is actually supposed to be used only by subclasses. In two years' time, I think I would just assume it's supposed to be accessible, since _that's what the code says_.

Also curious what @guozhangwang and @bbejeck think.

Reply:
Ah, I see you already added the interface. This is fine, too!

Reply:
The current hierarchy looks good to me.